File: sdks/python/src/osva/resources/workflows.py
from __future__ import annotations

from typing import Any

from osva.http_client import HTTPClient


class WorkflowsResource:
    def __init__(self, client: HTTPClient) -> None:
        self._client = client

    def list(self) -> dict[str, Any]:
        return self._client.request("GET", "/v1/workflows")

    def get(self, workflow_id: str) -> dict[str, Any]:
        return self._client.request("GET", f"/v1/workflows/{workflow_id}")

    def create(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self._client.request("POST", "/v1/workflows", body=payload)

    def create_version(self, workflow_id: str, payload: dict[str, Any]) -> dict[str, Any]:
        return self._client.request(
            "POST",
            f"/v1/workflows/{workflow_id}/versions",
            body=payload,
        )

File: packages/sdk/src/resources/workflows.ts
import type { WorkflowId, WorkflowVersionId } from "@osva/contracts";
import {
  createWorkflowRequestSchema,
  createWorkflowVersionRequestSchema,
  workflowListResourceSchema,
  workflowResourceSchema,
  workflowVersionListResourceSchema,
  workflowVersionResourceSchema,
} from "@osva/contracts/schemas";
import type { z } from "zod";

import type { OsvaHttpClient } from "../http-client.js";

type CreateWorkflowRequest = z.infer<typeof createWorkflowRequestSchema>;
type CreateWorkflowVersionRequest = z.infer<
  typeof createWorkflowVersionRequestSchema
>;
type WorkflowListResource = z.infer<typeof workflowListResourceSchema>;
type WorkflowResource = z.infer<typeof workflowResourceSchema>;
type WorkflowVersionListResource = z.infer<
  typeof workflowVersionListResourceSchema
>;
type WorkflowVersionResource = z.infer<typeof workflowVersionResourceSchema>;

export class WorkflowsResource {
  constructor(private readonly client: OsvaHttpClient) {}

  list(): Promise<WorkflowListResource> {
    return this.client.request({ method: "GET", path: "/v1/workflows" });
  }

  get(workflowId: WorkflowId): Promise<WorkflowResource> {
    return this.client.request({
      method: "GET",
      path: `/v1/workflows/${encodeURIComponent(workflowId)}`,
    });
  }

  create(input: CreateWorkflowRequest): Promise<WorkflowResource> {
    return this.client.request({
      method: "POST",
      path: "/v1/workflows",
      body: input,
    });
  }

  listVersions(workflowId: WorkflowId): Promise<WorkflowVersionListResource> {
    return this.client.request({
      method: "GET",
      path: `/v1/workflows/${encodeURIComponent(workflowId)}/versions`,
    });
  }

  getVersion(
    workflowId: WorkflowId,
    workflowVersionId: WorkflowVersionId,
  ): Promise<WorkflowVersionResource> {
    return this.client.request({
      method: "GET",
      path: `/v1/workflows/${encodeURIComponent(workflowId)}/versions/${encodeURIComponent(workflowVersionId)}`,
    });
  }

  createVersion(
    workflowId: WorkflowId,
    input: CreateWorkflowVersionRequest,
  ): Promise<WorkflowVersionResource> {
    return this.client.request({
      method: "POST",
      path: `/v1/workflows/${encodeURIComponent(workflowId)}/versions`,
      body: input,
    });
  }
}
